File: Engine/Android/core/src/test/kotlin/despia/engine/PasskeysConformanceTest.kt
package despia.engine

import java.io.File
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue
import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestFactory

/**
 * The SHARED PASSKEY corpus (OpenSource/Conformance/passkeys/ceremony.json) through THIS
 * runtime's Passkeys core — the Kotlin leg of Core/Passkeys (F17.3). The TS reference
 * (packages/kernel/test/passkeys-conformance.test.ts) and the Swift twin
 * (Engine/iOS/Passkeys.swift) read the SAME file, so a challenge cannot be padded on one
 * renderer and unpadded on another.
 */
class PasskeysConformanceTest {

    private fun corpusDir(): File {
        var dir = File(System.getProperty("user.dir") ?: ".").absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/passkeys")
            if (candidate.isDirectory) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/passkeys not found")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun doc(): Map<String, Any?> =
        json(File(corpusDir(), "ceremony.json").readText()).foundationValue as? Map<String, Any?>
            ?: error("ceremony.json: not a JSON object")

    @Suppress("UNCHECKED_CAST")
    private fun rows(key: String, minimum: Int): List<Map<String, Any?>> {
        val list = doc()[key] as? List<Map<String, Any?>> ?: error("no $key[]")
        assertTrue(list.size >= minimum, "passkeys/$key is suspiciously small (${list.size})")
        return list
    }

    private fun str(v: Any?): String = v as? String ?: ""
    private fun name(c: Map<String, Any?>): String = str(c["name"]).ifEmpty { "?" }

    @Suppress("UNCHECKED_CAST")
    private fun strings(v: Any?): List<String> = (v as? List<Any?> ?: emptyList()).map { str(it) }

    @Suppress("UNCHECKED_CAST")
    private fun ints(v: Any?): List<Int> = (v as? List<Any?> ?: emptyList()).map { (it as Number).toInt() }

    @Test
    @Suppress("UNCHECKED_CAST")
    fun limitsAndVocabulary() {
        val d = doc()
        val limits = d["limits"] as Map<String, Any?>
        assertEquals((limits["minChallengeBytes"] as Number).toInt(), Passkeys.MIN_CHALLENGE_BYTES)
        assertEquals((limits["maxChallengeBytes"] as Number).toInt(), Passkeys.MAX_CHALLENGE_BYTES)
        assertEquals((limits["maxUserIdBytes"] as Number).toInt(), Passkeys.MAX_USER_ID_BYTES)
        assertEquals((limits["minTimeoutMs"] as Number).toInt(), Passkeys.MIN_TIMEOUT_MS)
        assertEquals((limits["defaultTimeoutMs"] as Number).toInt(), Passkeys.DEFAULT_TIMEOUT_MS)
        assertEquals((limits["maxTimeoutMs"] as Number).toInt(), Passkeys.MAX_TIMEOUT_MS)
        val vocab = d["vocabulary"] as Map<String, Any?>
        assertEquals(strings(vocab["attestations"]), Passkeys.ATTESTATIONS)
        assertEquals(strings(vocab["mediations"]), Passkeys.MEDIATIONS)
        assertEquals(strings(vocab["verifications"]), Passkeys.VERIFICATIONS)
        assertEquals(strings(vocab["residentKeys"]), Passkeys.RESIDENT_KEYS)
    }

    @TestFactory
    fun encode(): List<DynamicTest> = rows("encode", 8).map { c ->
        DynamicTest.dynamicTest("passkeys/encode — ${name(c)}") {
            val bytes = ints(c["bytes"])
            assertEquals(str(c["expect"]), Passkeys.base64UrlEncode(bytes))
            // Every encode case is also a round trip: the codec's only real contract.
            assertEquals(bytes, Passkeys.base64UrlDecode(str(c["expect"])))
        }
    }

    @TestFactory
    fun decode(): List<DynamicTest> = rows("decode", 8).map { c ->
        DynamicTest.dynamicTest("passkeys/decode — ${name(c)}") {
            val got = Passkeys.base64UrlDecode(c["text"])
            if (c["ok"] == true) assertEquals(ints(c["bytes"]), got) else assertNull(got)
        }
    }

    private fun <T> check(got: Result<T>, c: Map<String, Any?>, assertValue: (T) -> Unit) {
        assertEquals(c["ok"] == true, got.isSuccess, "ok")
        if (c["ok"] == true) {
            assertValue(got.getOrThrow())
        } else {
            assertEquals(str(c["error"]), Passkeys.code(got.exceptionOrNull()!!), "error")
        }
    }

    @TestFactory
    fun rpId(): List<DynamicTest> = rows("rpId", 12).map { c ->
        DynamicTest.dynamicTest("passkeys/rpId — ${name(c)}") {
            check(Passkeys.normalizeRpId(c["raw"]), c) { assertEquals(str(c["expect"]), it) }
        }
    }

    @TestFactory
    fun challenge(): List<DynamicTest> = rows("challenge", 6).map { c ->
        DynamicTest.dynamicTest("passkeys/challenge — ${name(c)}") {
            check(Passkeys.normalizeChallenge(c["raw"]), c) { assertEquals(str(c["expect"]), it) }
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun userOf(v: Any?): Passkeys.User {
        val m = v as Map<String, Any?>
        return Passkeys.User(str(m["id"]), str(m["name"]), str(m["displayName"]))
    }

    @TestFactory
    fun user(): List<DynamicTest> = rows("user", 8).map { c ->
        DynamicTest.dynamicTest("passkeys/user — ${name(c)}") {
            check(Passkeys.normalizeUser(c["raw"]), c) { assertEquals(userOf(c["expect"]), it) }
        }
    }

    @TestFactory
    fun words(): List<DynamicTest> = rows("words", 10).map { c ->
        DynamicTest.dynamicTest("passkeys/words — ${name(c)}") {
            val got = when (str(c["kind"])) {
                "attestation" -> Passkeys.foldAttestation(c["raw"])
                "mediation" -> Passkeys.foldMediation(c["raw"])
                "verification" -> Passkeys.foldUserVerification(c["raw"])
                "residentKey" -> Passkeys.foldResidentKey(c["raw"])
                else -> error("no fold for ${str(c["kind"])}")
            }
            check(got, c) { assertEquals(str(c["expect"]), it) }
        }
    }

    @TestFactory
    fun timeout(): List<DynamicTest> = rows("timeout", 5).map { c ->
        DynamicTest.dynamicTest("passkeys/timeout — ${name(c)}") {
            assertEquals((c["expect"] as Number).toInt(), Passkeys.clampTimeout(c["raw"]))
        }
    }

    @TestFactory
    fun credentials(): List<DynamicTest> = rows("credentials", 6).map { c ->
        DynamicTest.dynamicTest("passkeys/credentials — ${name(c)}") {
            check(Passkeys.normalizeCredentialIds(c["raw"]), c) { assertEquals(strings(c["expect"]), it) }
        }
    }

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun create(): List<DynamicTest> = rows("create", 5).map { c ->
        DynamicTest.dynamicTest("passkeys/create — ${name(c)}") {
            check(Passkeys.normalizeCreateOptions(c["raw"] as Map<String, Any?>), c) { got ->
                val want = c["expect"] as Map<String, Any?>
                assertEquals(str(want["rpId"]), got.rpId, "rpId")
                assertEquals(userOf(want["user"]), got.user, "user")
                assertEquals(str(want["challenge"]), got.challenge, "challenge")
                assertEquals(str(want["attestation"]), got.attestation, "attestation")
                assertEquals(str(want["userVerification"]), got.userVerification, "userVerification")
                assertEquals(str(want["residentKey"]), got.residentKey, "residentKey")
                assertEquals(strings(want["excludeCredentials"]), got.excludeCredentials, "exclude")
                assertEquals((want["timeoutMs"] as Number).toInt(), got.timeoutMs, "timeoutMs")
            }
        }
    }

    @TestFactory
    @Suppress("UNCHECKED_CAST")
    fun get(): List<DynamicTest> = rows("get", 4).map { c ->
        DynamicTest.dynamicTest("passkeys/get — ${name(c)}") {
            check(Passkeys.normalizeGetOptions(c["raw"] as Map<String, Any?>), c) { got ->
                val want = c["expect"] as Map<String, Any?>
                assertEquals(str(want["rpId"]), got.rpId, "rpId")
                assertEquals(str(want["challenge"]), got.challenge, "challenge")
                assertEquals(str(want["mediation"]), got.mediation, "mediation")
                assertEquals(str(want["userVerification"]), got.userVerification, "userVerification")
                assertEquals(strings(want["allowCredentials"]), got.allowCredentials, "allow")
                assertEquals((want["timeoutMs"] as Number).toInt(), got.timeoutMs, "timeoutMs")
            }
        }
    }
}

File: Web/packages/kernel/test/passkeys-conformance.test.ts
//
//  passkeys-conformance.test.ts — the SHARED PASSKEY corpus
//  (OpenSource/Conformance/passkeys/ceremony.json) through the TS core, the REFERENCE leg of
//  Core/Passkeys (F17.3). The Kotlin twin (:core PasskeysConformanceTest) and the Swift twin
//  (Engine/iOS/Passkeys.swift) read the SAME file, so a challenge cannot be padded on one
//  renderer and unpadded on another.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  PASSKEY_MIN_CHALLENGE_BYTES, PASSKEY_MAX_CHALLENGE_BYTES, PASSKEY_MAX_USER_ID_BYTES,
  PASSKEY_MIN_TIMEOUT_MS, PASSKEY_DEFAULT_TIMEOUT_MS, PASSKEY_MAX_TIMEOUT_MS,
  PASSKEY_ATTESTATIONS, PASSKEY_MEDIATIONS, PASSKEY_VERIFICATIONS, PASSKEY_RESIDENT_KEYS,
  base64UrlEncode, base64UrlDecode, normalizeRpId, normalizeChallenge, normalizeUser,
  foldAttestation, foldMediation, foldUserVerification, foldResidentKey,
  clampPasskeyTimeout, normalizeCredentialIds, normalizeCreateOptions, normalizeGetOptions,
  type PasskeyResult,
} from "../src/passkeys.ts";

function corpusDir(): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, "OpenSource/Conformance/passkeys");
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error("passkeys corpus not found");
    dir = parent;
  }
}

type Doc = {
  limits: Record<string, number>;
  vocabulary: { attestations: string[]; mediations: string[]; verifications: string[]; residentKeys: string[] };
  encode: { name: string; bytes: number[]; expect: string }[];
  decode: { name: string; text: string; ok: boolean; bytes?: number[] }[];
  rpId: { name: string; raw: string; ok: boolean; expect?: string; error?: string }[];
  challenge: { name: string; raw: string; ok: boolean; expect?: string; error?: string }[];
  user: { name: string; raw: unknown; ok: boolean; expect?: unknown; error?: string }[];
  words: { name: string; kind: string; raw: unknown; ok: boolean; expect?: string; error?: string }[];
  timeout: { name: string; raw: unknown; expect: number }[];
  credentials: { name: string; raw: unknown; ok: boolean; expect?: string[]; error?: string }[];
  create: { name: string; raw: Record<string, unknown>; ok: boolean; expect?: unknown; error?: string }[];
  get: { name: string; raw: Record<string, unknown>; ok: boolean; expect?: unknown; error?: string }[];
};

const doc = JSON.parse(readFileSync(join(corpusDir(), "ceremony.json"), "utf-8")) as Doc;

assert.ok(doc.encode.length >= 8, "passkeys/encode corpus is suspiciously small");
assert.ok(doc.rpId.length >= 12, "passkeys/rpId corpus is suspiciously small");
assert.ok(doc.create.length >= 5 && doc.get.length >= 4, "passkeys ceremony corpus is suspiciously small");

test("passkeys — the pinned limits and vocabulary", () => {
  assert.equal(PASSKEY_MIN_CHALLENGE_BYTES, doc.limits["minChallengeBytes"]);
  assert.equal(PASSKEY_MAX_CHALLENGE_BYTES, doc.limits["maxChallengeBytes"]);
  assert.equal(PASSKEY_MAX_USER_ID_BYTES, doc.limits["maxUserIdBytes"]);
  assert.equal(PASSKEY_MIN_TIMEOUT_MS, doc.limits["minTimeoutMs"]);
  assert.equal(PASSKEY_DEFAULT_TIMEOUT_MS, doc.limits["defaultTimeoutMs"]);
  assert.equal(PASSKEY_MAX_TIMEOUT_MS, doc.limits["maxTimeoutMs"]);
  assert.deepEqual([...PASSKEY_ATTESTATIONS], doc.vocabulary.attestations);
  assert.deepEqual([...PASSKEY_MEDIATIONS], doc.vocabulary.mediations);
  assert.deepEqual([...PASSKEY_VERIFICATIONS], doc.vocabulary.verifications);
  assert.deepEqual([...PASSKEY_RESIDENT_KEYS], doc.vocabulary.residentKeys);
});

for (const c of doc.encode) {
  test(`passkeys/encode — ${c.name}`, () => {
    assert.equal(base64UrlEncode(c.bytes), c.expect);
    // Every encode case is also a round trip: the codec's only real contract.
    assert.deepEqual(base64UrlDecode(c.expect), c.bytes);
  });
}

for (const c of doc.decode) {
  test(`passkeys/decode — ${c.name}`, () => {
    const got = base64UrlDecode(c.text);
    if (c.ok) assert.deepEqual(got, c.bytes);
    else assert.equal(got, null);
  });
}

function check<T>(got: PasskeyResult<T>, c: { ok: boolean; error?: string }, expect: unknown): void {
  assert.equal(got.ok, c.ok, got.ok ? "ok" : `ok (${got.detail ?? ""})`);
  if (got.ok) assert.deepEqual(JSON.parse(JSON.stringify(got.value)), expect);
  else assert.equal(got.error, c.error, "error");
}

for (const c of doc.rpId) {
  test(`passkeys/rpId — ${c.name}`, () => check(normalizeRpId(c.raw), c, c.expect));
}

for (const c of doc.challenge) {
  test(`passkeys/challenge — ${c.name}`, () => check(normalizeChallenge(c.raw), c, c.expect));
}

for (const c of doc.user) {
  test(`passkeys/user — ${c.name}`, () => check(normalizeUser(c.raw), c, c.expect));
}

const FOLDS: Record<string, (raw: unknown) => PasskeyResult<string>> = {
  attestation: foldAttestation,
  mediation: foldMediation,
  verification: foldUserVerification,
  residentKey: foldResidentKey,
};

for (const c of doc.words) {
  test(`passkeys/words — ${c.name}`, () => {
    const fold = FOLDS[c.kind];
    assert.ok(fold !== undefined, `no fold for ${c.kind}`);
    check(fold(c.raw), c, c.expect);
  });
}

for (const c of doc.timeout) {
  test(`passkeys/timeout — ${c.name}`, () => {
    assert.equal(clampPasskeyTimeout(c.raw), c.expect);
  });
}

for (const c of doc.credentials) {
  test(`passkeys/credentials — ${c.name}`, () => check(normalizeCredentialIds(c.raw), c, c.expect));
}

for (const c of doc.create) {
  test(`passkeys/create — ${c.name}`, () => check(normalizeCreateOptions(c.raw), c, c.expect));
}

for (const c of doc.get) {
  test(`passkeys/get — ${c.name}`, () => check(normalizeGetOptions(c.raw), c, c.expect));
}
